middleware.go

```diff
@@ -0,0 +1,107 @@
+package jwt
+
+import (
+	"errors"
+	"github.com/Digital-Insight-Technologies-Ltd/gin-jwt-cognito/models"
+	"github.com/sirupsen/logrus"
+	"net/http"
+	"strings"
+
+	"github.com/gin-gonic/gin"
+	"github.com/golang-jwt/jwt/v5"
+	"github.com/google/uuid"
+)
+
+// UserClaimsKey is the key for user claims in our context
+const UserClaimsKey = "userClaims"
+
+// CognitoClaimsMiddleware is a middleware for Gin that extracts user claims from
+// AWS Cognito.
+func CognitoClaimsMiddleware(jwtParser *jwt.Parser) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		authHeader := c.GetHeader("Authorization")
+		if !strings.HasPrefix(authHeader, "Bearer ") {
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Missing or invalid Authorization header"})
+			return
+		}
+		claims, err := parseClaims(jwtParser, authHeader[7:])
+
+		if err != nil {
+			logrus.WithError(err).Error("Error parsing claims (OIDC: stage 1)")
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 1)"})
+			return
+		}
+
+		userCtx, err := constructUserContext(claims)
+
+		if err != nil {
+			logrus.WithError(err).Error("Error constructing user context (OIDC: stage 2)")
+			c.AbortWithStatusJSON(http.StatusUnauthorized, gin.H{"error": "Invalid user claims (OIDC: stage 2)"})
+			return
+		}
+
+		// Add the claims to Gin's context
+		c.Set(UserClaimsKey, userCtx)
+
+		// Proceed with the middleware chain
+		c.Next()
+	}
+}
+
+func parseClaims(jwtParser *jwt.Parser, claimsHeader string) (jwt.MapClaims, error) {
+	tok, _, err := jwtParser.ParseUnverified(claimsHeader, jwt.MapClaims{})
```

Comment:
In `parseClaims`, ensure that the JWT token's signature is validated to prevent accepting tampered or unsigned tokens. Use `jwtParser.Parse` instead of `ParseUnverified` and provide a key function to verify the signature. [important]